### PR TITLE
split ext_authz fuzzer into http & grpc fuzzers

### PR DESCRIPTION
**Pull request by antoniovleonti**
Commit Message: split ext_authz fuzzer into http & grpc fuzzers
Additional Description: This PR does the following:
1. create a library for fuzzing ext authz that takes a config and a client and returns a filter used for fuzzing.
2. modify the ext_authz grpc fuzzer test to use this library
3. create a new ext_authz http fuzzer test that also uses the library

Risk Level: none
Testing: increases fuzzing coverage for ext_authz (specifically for the http client)
Docs Changes: none
Release Notes: none
Platform Specific Features: none
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by yanjunxiang-google** [comment]
Thanks for working on this!  The change looks good to me from high level.

File: test/extensions/filters/http/ext_authz/ext_authz_grpc_fuzz_test.cc [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+#include "test/extensions/filters/common/ext_authz/mocks.h"
+#include "test/extensions/filters/common/ext_authz/test_common.h"
+#include "test/extensions/filters/http/common/fuzz/http_filter_fuzzer.h"
+#include "test/extensions/filters/http/ext_authz/ext_authz_fuzz.pb.validate.h"
+#include "test/extensions/filters/http/ext_authz/ext_authz_fuzz_lib.h"
+#include "test/fuzz/fuzz_runner.h"
+#include "test/mocks/grpc/mocks.h"
+#include "test/mocks/network/mocks.h"
+#include "test/mocks/server/server_factory_context.h"
```
Comment (yanjunxiang-google) [inline test/extensions/filters/http/ext_authz/ext_authz_grpc_fuzz_test.cc:15]: nit: are we able to clean up some of the included .h files?

File: test/extensions/filters/http/ext_authz/ext_authz_grpc_fuzz_test.cc [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+using testing::Return;
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace ExtAuthz {
+namespace {
+
+std::unique_ptr<envoy::service::auth::v3::CheckResponse>
+makeGrpcCheckResponse(const Grpc::Status::WellKnownGrpcStatus status) {
```
Comment (yanjunxiang-google) [inline test/extensions/filters/http/ext_authz/ext_authz_grpc_fuzz_test.cc:30]: Instead of writing code to make CheckResponse, we should change ExtAuthzTest proto
https://github.com/envoyproxy/envoy/blob/412a936100888530cfa1746621a933e78edc78af/test/extensions/filters/http/ext_authz/ext_authz_fuzz.proto#L10

to remove the AuthResult from there, 
https://github.com/envoyproxy/envoy/blob/412a936100888530cfa1746621a933e78edc78af/test/extensions/filters/http/ext_authz/ext_authz_fuzz.proto#L24

instead, adding a CheckResponse message there:
https://github.com/envoyproxy/envoy/blob/412a936100888530cfa1746621a933e78edc78af/api/envoy/service/auth/v3/external_auth.proto#L117

That way, the fuzzer will be able to generate random CheckResponse and feed into the test input:
https://github.com/envoyproxy/envoy/blob/412a936100888530cfa1746621a933e78edc78af/test/extensions/filters/http/ext_authz/ext_authz_fuzz_test.cc#L78

Comment (yanjunxiang-google) [reply test/extensions/filters/http/ext_authz/ext_authz_grpc_fuzz_test.cc:30]: A good example like this is the ext_proc fuzzer test input:
https://github.com/envoyproxy/envoy/blob/412a936100888530cfa1746621a933e78edc78af/test/extensions/filters/http/ext_proc/unit_test_fuzz/ext_proc_unit_test_fuzz.proto#L11

File: test/extensions/filters/http/ext_authz/ext_authz_grpc_fuzz_test.cc [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+  }
+  case envoy::extensions::filters::http::ext_authz::ExtAuthzTestCase::ERROR: {
+    check_status = Grpc::Status::WellKnownGrpcStatus::Internal;
+    break;
+  }
+  case envoy::extensions::filters::http::ext_authz::ExtAuthzTestCase::DENIED: {
+    check_status = Grpc::Status::WellKnownGrpcStatus::PermissionDenied;
+    break;
+  }
+  default: {
```
Comment (tyxia) [inline test/extensions/filters/http/ext_authz/ext_authz_grpc_fuzz_test.cc:54]: Thanks for working on this!
Just a quick drive-by comment. If you do PANIC here, you probably want to remove default case, and use exhaustive switch case.  PANIC_ON_PROTO_ENUM_SENTINEL_VALUES/PANIC_DUE_TO_CORRUPT_ENUM

File: test/extensions/filters/http/ext_authz/ext_authz_grpc_fuzz_test.cc (hunk 2 above)
Comment (antoniovleonti) [reply test/extensions/filters/http/ext_authz/ext_authz_grpc_fuzz_test.cc:30]: This is just copied [directly from the old fuzzer code](https://github.com/envoyproxy/envoy/pull/33012/files#diff-de8edc70c81913c4a3089c1e123dd67d2d53860eb3deaa2b6d83e3acc90390cbL28). I will add CheckResponse to the fuzzer input in a subsequent PR.

File: test/extensions/filters/http/ext_authz/ext_authz_grpc_fuzz_test.cc (hunk 1 above)
Comment (antoniovleonti) [reply test/extensions/filters/http/ext_authz/ext_authz_grpc_fuzz_test.cc]: Yes, done.

File: test/extensions/filters/http/ext_authz/ext_authz_grpc_fuzz_test.cc (hunk 3 above)
Comment (antoniovleonti) [reply test/extensions/filters/http/ext_authz/ext_authz_grpc_fuzz_test.cc]: TIL! Done.

**Comment by yanavlasov** [comment]
@antoniovleonti please resolve merge conflict

@tyxia @yanjunxiang-google please take another look after merge.

**Comment by antoniovleonti** [comment]
Sorry about the reviewer spam, I accidentally rebased :-(

**Comment by yanjunxiang-google** [comment]
Could you please take a look the ASAN error?

File: test/extensions/filters/http/ext_authz/BUILD [hunk 4]
```diff
@@ -98,21 +99,56 @@ envoy_proto_library(
...
+        "@envoy_api//envoy/service/auth/v3:pkg_cc_proto",
+    ],
+)
+
+envoy_cc_fuzz_test(
+    name = "ext_authz_http_fuzz_test",
+    srcs = ["ext_authz_http_fuzz_test.cc"],
+    corpus = "ext_authz_corpus",
+    deps = [
+        ":ext_authz_fuzz_lib",
```
Comment (yanjunxiang-google) [inline test/extensions/filters/http/ext_authz/BUILD:124]: As you put a lot of dependencies in your ext_authz_fuzz_lib already, do you need to directly include them again, like 

"//source/extensions/filters/http/ext_authz",
"//test/extensions/filters/common/ext_authz:ext_authz_mocks",
...

I would rather just include ext_authz_fuzz_lib and other dependencies which is not included in ext_authz_fuzz_lib. And let the compiler bring in those indirect dependencies to keep the BUILD target simple and clean.  Same as above.

Comment (antoniovleonti) [reply test/extensions/filters/http/ext_authz/BUILD:122]: I found a couple I wasn't using directly in the http and grpc fuzz tests and removed them.

File: test/extensions/filters/http/ext_authz/ext_authz_fuzz_lib.cc [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    return absl::InvalidArgumentError(absl::StrCat("EnvoyException during validation: ", e.what()));
+  }
+
+  // Prepare filter.
+  const envoy::extensions::filters::http::ext_authz::v3::ExtAuthz proto_config = input.config();
+  FilterConfigSharedPtr config;
+
+  try {
+    config = std::make_shared<FilterConfig>(proto_config, *stats_store_.rootScope(),
+                                            "ext_authz_prefix", factory_context_);
```
Comment (yanjunxiang-google) [inline test/extensions/filters/http/ext_authz/ext_authz_fuzz_lib.cc:61]: Should "factory_context_" actually be a type envoy::config::bootstrap::v3::Bootstrap.  Just curious how this compiles?

File: test/extensions/filters/http/ext_authz/ext_authz_grpc_fuzz_test.cc [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+              // TODO: Query the request header map in HttpFilterFuzzer to test
+              // headers_to_(add/remove/append).
+              // TODO: Test check request attributes against config
+              // and filter metadata.
+              ENVOY_LOG_MISC(trace, "Check Request attributes {}",
+                             check_request.attributes().DebugString());
+
+              if (status_ == Grpc::Status::WellKnownGrpcStatus::Ok) {
+                grpc_client_->onSuccess(makeGrpcCheckResponse(status_), mock_span_);
```
Comment (yanjunxiang-google) [inline test/extensions/filters/http/ext_authz/ext_authz_grpc_fuzz_test.cc:70]: Good job to mock the response from client layer, instead of directly call filter layer code.  This will expand the fuzzer code coverage to ext_authz gRPC client code.  And same for ext_authz HTTP client code.

File: test/extensions/filters/http/ext_authz/ext_authz_http_fuzz_test.cc [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+  NiceMock<Server::Configuration::MockServerFactoryContext> factory_context_;
+
+  // Set by calling newRawHttpClientImpl
+  std::string status_;
+  Filters::Common::ExtAuthz::RawHttpClientImpl* http_client_;
+};
+
+DEFINE_PROTO_FUZZER(const ExtAuthzTestCase& input) {
+  static ReusableFuzzerUtil fuzzer_util;
+  static ReusableHttpClientFactory http_client_factory;
```
Comment (yanjunxiang-google) [inline test/extensions/filters/http/ext_authz/ext_authz_http_fuzz_test.cc:126]: I am wondering what's the benefit to have a factory to create the filter instead of directly call a library function to create the filter, like you can put the code from https://github.com/envoyproxy/envoy/blob/08e23c4eb31ad0a642f415c8872d542364083b34/test/extensions/filters/http/ext_authz/ext_authz_fuzz_test.cc#L93 to https://github.com/envoyproxy/envoy/blob/08e23c4eb31ad0a642f415c8872d542364083b34/test/extensions/filters/http/ext_authz/ext_authz_fuzz_test.cc#L104 into the library function and call it.

**Comment by yanavlasov** [comment]
Waiting for comments to be resolved.

/wait-any

File: test/extensions/filters/http/ext_authz/ext_authz_fuzz_lib.cc (hunk 5 above)
Comment (antoniovleonti) [reply test/extensions/filters/http/ext_authz/ext_authz_fuzz_lib.cc:61]: This is just copied from [the current fuzzer test](https://github.com/envoyproxy/envoy/blob/22ad8865ef54ccaa239108e0154aedae0f994aab/test/extensions/filters/http/ext_authz/ext_authz_fuzz_test.cc#L94
).

File: test/extensions/filters/http/ext_authz/ext_authz_http_fuzz_test.cc (hunk 7 above)
Comment (antoniovleonti) [reply test/extensions/filters/http/ext_authz/ext_authz_http_fuzz_test.cc:126]: The factory makes it easier to run individual fuzzer test cases without recreating all the mocks each time. The mock call expectations just get set once and the data they return gets updated with each test call. Notice the fuzzer util is static in the actual fuzzer tests so it get reused.

Another benefit is that all the code that is shared between the grpc and http fuzzers is in this file instead of copied between the two new fuzzers.

**Comment by yanjunxiang-google** [comment]
LGTM

**Review by yanjunxiang-google — APPROVED** [review]

**Review by yanavlasov — APPROVED** [review]
